Add support for using Registry's via tuple

acco opened the pull request:
Support overriding process_name/2 in Broadway module so that all processes in topology can be started with a via tuple.

Discussed in #236

josevalim commented on lib/broadway/topology/subscriber.ex:
```diff
@@ -133,6 +133,14 @@ defmodule Broadway.Topology.Subscriber do
     end
   end
 
+  defp get_pid(process_name) when is_atom(process_name) do
+    Process.whereis(process_name)
+  end
+
+  defp get_pid({:via, Registry, {registry, args}}) do
+    registry.whereis_name(args)
+  end
```
Use `GenServer.whereis/1` instead. :)

acco replied:
Wow!

josevalim commented on test/broadway_test.exs:
```diff
@@ -487,6 +481,69 @@ defmodule BroadwayTest do
+        def whereis_name(name) do
+          Registry.whereis_name({__MODULE__, name})
+        end
+      end
```
I believe we don't need to create this module, instead, just use `Registry` directly?

josevalim commented on test/broadway_test.exs:
```diff
@@ -487,6 +481,69 @@ defmodule BroadwayTest do
+
+        def handle_message(_, _, _), do: nil
+        def handle_batch(_, _, _, _), do: nil
+      end
```
I believe we can instead use any of the other Broadway pipelines in this module. :)

josevalim commented on test/broadway_test.exs:
```diff
@@ -487,6 +481,69 @@ defmodule BroadwayTest do
+        def handle_batch(_, _, _, _), do: nil
+      end
+
+      defmodule SomeBroadway do
```
`RegistryBroadway`?

josevalim commented:
Great work! I have added some minor comments.

We also need to add tests for Broadway.get_rate_limiting, Broadway.test_messages, Broadway.producer_names, and Broadway.topology. So I think we should move the test that we have so far into its own describe block, so we can add these new tests. WDYT?

acco commented on lib/broadway/topology.ex:
```diff
@@ -151,12 +151,16 @@ defmodule Broadway.Topology do
+  end
+
+  defp put_terminator(config) do
+    Map.put(config, :terminator, process_name(config, "Terminator"))
```
We need to wait until `config` is built before calling `process_name`, which takes the config.

The alternative is to change `process_name` from this:

```
process_name(config, "Terminator")
```

To this:

```
process_name(config.module, config.name, "Terminator"
```

Where the module and broadway name are extracted from the config first then passed down to `process_name`. Happy to go that route if desired

acco replied:
@josevalim You move fast! 😂 While you were reviewing, I switched from `process_name(config.module, config.name, "Terminator")` -> `process_name(config, "Terminator")`, but happy to switch back

josevalim replied:
I like `process_name/2` too.

acco commented on lib/broadway/topology/rate_limiter.ex:
```diff
@@ -107,4 +107,12 @@ defmodule Broadway.Topology.RateLimiter do
+
+  defp get_pid({:via, Registry, {registry, args}}) do
+    registry.whereis_name(args)
+  end
```
Not sure if there's a cleaner way to do this – can't rely on `Process.whereis` to find everything any more

acco replied:
Saw your comment - TIL!

josevalim commented on lib/broadway/options.ex:
```diff
@@ -5,7 +5,7 @@ defmodule Broadway.Options do
+        type: :any,
         doc: """
         Used for name registration. All processes/stages
         created will be named using this value as prefix.
```
You can provide a custom MFA for validation. You should likely validate it is an atom or a via tuple.

acco commented:
Sounds good @josevalim ! Updates coming, thanks so much for your quick feedback

acco commented on lib/broadway.ex:
```diff
@@ -897,8 +912,8 @@ defmodule Broadway do
-  @spec producer_names(broadway :: atom()) :: [atom()]
-  def producer_names(broadway) when is_atom(broadway) do
+  @spec producer_names(broadway :: atom() | via_tuple()) :: [atom()]
+  def producer_names(broadway) do
```
Had to lose the guard clause for `is_atom()` with these functions. Lmk if there's a better option?

whatyouhide replied:
You can do

```suggestion
  def producer_names(broadway) when is_atom(broadway) or (is_tuple(broadway) and tuple_size(broadway) == 3) do
```

and it already catches most cases of wrong stuff going in, IMO :)

whatyouhide replied:
We can add a private guard for this maybe 

```elixir
defguardp is_broadway_name(...) ...
```

josevalim replied:
I like the guard route!

acco replied:
TIL!

acco commented on lib/broadway/topology.ex:
```diff
@@ -151,12 +150,24 @@ defmodule Broadway.Topology do
       resubscribe_interval: opts[:resubscribe_interval]
     }
+    |> put_terminator()
+    |> put_rate_limiter(opts)
```
Moved `rate_limiter` (its name) into config. This lines it up with eg Terminator and ensures we don't have RateLimiter name itself (as you'll see below)

Lmk if this works

josevalim replied:
This works but have both `terminator: nil` and `rate_limiter: nil` in the initial map and simply update the keys instead. :)

acco replied:
Good call

acco commented on lib/broadway/topology.ex:
```diff
@@ -215,11 +226,11 @@ defmodule Broadway.Topology do
       end
 
-    args = [broadway: opts, dispatcher: dispatcher] ++ producer_config
+    args = [broadway: opts, dispatcher: dispatcher, rate_limiter: rate_limiter] ++ producer_config
```
Pass the name for the rate limiter down through ProducerStage

acco commented on lib/broadway/topology/producer_stage.ex:
```diff
@@ -44,16 +43,13 @@ defmodule Broadway.Topology.ProducerStage do
-          |> RateLimiter.rate_limiter_name()
-          |> RateLimiter.get_rate_limiter_ref()
+      if rate_limiter do
+        rate_limiter_ref = RateLimiter.get_rate_limiter_ref(rate_limiter)
```
I guess we could also just pass down `rate_limiter_ref` to ProducerStage

acco commented on lib/broadway/topology/rate_limiter.ex:
```diff
@@ -15,8 +15,8 @@ defmodule Broadway.Topology.RateLimiter do
-        args = {name, rate_limiting_opts, producers_names}
-        GenServer.start_link(__MODULE__, args, name: rate_limiter_name(name))
+        args = {rate_limiting_opts, producers_names}
+        GenServer.start_link(__MODULE__, args, name: name)
```
Name is now derived externally

acco commented on test/broadway_test.exs:
```diff
@@ -2522,7 +2611,7 @@ defmodule BroadwayTest do
 
   defp get_rate_limiter(broadway_name) do
-    :"#{broadway_name}.RateLimiter"
+    :"#{broadway_name}.Broadway.RateLimiter"
```
Respect namespace that all processes have via `process_name/2`

acco reviewed (commented):
Alright @josevalim see updates!

Assuming we're getting close, let me know where in docs I can make updates

josevalim commented on lib/broadway.ex:
```diff
@@ -681,6 +681,7 @@ defmodule Broadway do
   Returned by `start_link/2`.
   """
   @type on_start() :: {:ok, pid()} | :ignore | {:error, {:already_started, pid()} | term()}
+  @type via_tuple :: {:via, Registry, any()}
```
```suggestion
  @type via_tuple :: {:via, module, any()}
```

josevalim replied:
Or maybe even better:

```suggestion
  @type name :: atom() | {:via, module(), term()}
```

And then we can reuse it throughout the API?

josevalim commented on lib/broadway.ex:
```diff
@@ -830,7 +831,21 @@ defmodule Broadway do
       end
 
-      defoverridable child_spec: 1
+      @type via_tuple :: {:via, Registry, any()}
+      @callback process_name(broadway_name :: String.t() | via_tuple, base_name :: String.t()) ::
+                  any()
```
```suggestion
      @callback process_name(broadway_name :: atom() | {:via, module(), term()} base_name :: String.t()) ::
                  any()
```

josevalim commented on lib/broadway/options.ex:
```diff
@@ -290,4 +290,12 @@ defmodule Broadway.Options do
+
+  def validate_name(name) when is_atom(name), do: {:ok, name}
+
+  def validate_name({:via, Registry, {_, _}} = via), do: {:ok, via}
```
```suggestion
  def validate_name({:via, _, _} = via), do: {:ok, via}
```

josevalim commented on lib/broadway/options.ex:
```diff
@@ -290,4 +290,12 @@ defmodule Broadway.Options do
+  def validate_name({:via, Registry, {_, _}} = via), do: {:ok, via}
+
+  def validate_name(name) do
+    {:error, "expected :name to be an atom or a via tuple, got: #{inspect(name)}"}
```
```suggestion
    {:error, "expected :name to be an atom or a {:via, module, term} tuple, got: #{inspect(name)}"}
```

josevalim commented on lib/broadway/topology.ex:
```diff
@@ -151,12 +150,24 @@ defmodule Broadway.Topology do
+  end
+
+  defp put_rate_limiter(config, opts) do
+    case get_in(opts, [:producer, :rate_limiting]) do
+      nil -> Map.put(config, :rate_limiter, nil)
+      _ -> Map.put(config, :rate_limiter, process_name(config, "RateLimiter"))
+    end
```
If you initialize it to nil above, then:

```suggestion
    if opts[:producer][:rate_limiting] do
      Map.put(config, :rate_limiter, process_name(config, "RateLimiter"))
    else
      config
    end
```

josevalim commented on test/broadway_test.exs:
```diff
@@ -2465,6 +2476,84 @@ defmodule BroadwayTest do
+  describe "starting under a registry using via tuple" do
+    setup do
+      {:ok, _registry} = start_supervised({Registry, keys: :unique, name: MyRegistry})
+      name = via_tuple({UsesRegistry, "1"})
```
For simplicity, I would have this simply be:

    via_tuple(:broadway)

We probably don't need a tuple. :)

acco replied:
Good call

josevalim commented:
@acco fantastic! I have dropped the last batch of comments. Then please rebase and we should be good to go!

acco commented:
Thanks @josevalim, great notes! Just incorporated, rebased, and pushed. Assuming latest looks good, I can squash and we can merge. Let me know if it needs any other tweaks

whatyouhide commented on lib/broadway/options.ex:
```diff
@@ -290,4 +290,13 @@ defmodule Broadway.Options do
+
+  def validate_name(name) when is_atom(name), do: {:ok, name}
+
+  def validate_name({:via, _, _} = via), do: {:ok, via}
```
Maybe

```suggestion
  def validate_name({:via, module, _term} = via) when is_atom(module), do: {:ok, via}
```

josevalim commented on lib/broadway.ex:
```diff
@@ -859,7 +860,21 @@ defmodule Broadway do
       end
 
-      defoverridable child_spec: 1
+      @type name :: atom() | {:via, module(), term()}
+      @callback process_name(broadway_name :: name(), base_name :: String.t()) ::
+                  any()
```
```suggestion
      @callback process_name(Broadway.name(), base_name :: String.t()) :: Broadway.name()
```

josevalim commented on lib/broadway.ex:
```diff
@@ -926,8 +941,8 @@ defmodule Broadway do
   """
-  @spec producer_names(broadway :: atom()) :: [atom()]
-  def producer_names(broadway) when is_atom(broadway) do
+  @spec producer_names(broadway :: name()) :: [atom()]
```
```suggestion
  @spec producer_names(name()) :: [name()]
```

josevalim commented:
@acco I added some very minor comments and @whatyouhide suggested a guard, which can be very helpful, and we can ship it!

acco commented:
Great feedback/tweets @josevalim @whatyouhide , thank you! Pushed latest. Happy to squash before merge, lmk

josevalim commented:
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

acco commented:
Wooo! Thanks for all your help 🙏
